Implement Worker.prototype.terminate()

jwilm opened the pull request:
Adds support for terminating DOM workers. `ScriptMsg::Terminate` was added to notify the WorkerGlobalScope/DedicatedWorkerGlobalScope to stop processing events and shut down. A closing flag was added to WorkerGlobalScope per the spec.

Additionally, a closing flag was added to `dom::worker::Worker` to ensure no further events are dispatched to the `onmessage` handler.

The change to the wpt WorkerTerminate.js is a patch for the broken Worker_terminate_event_queue test (asserts 10001 == 10000).

Resolves #4427

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5846)

<!-- Reviewable:end -->

jwilm committed:
Implement Worker.prototype.terminate()

Adds support for terminating DOM workers. A new ScriptMsg::Terminate was
added to notify the WorkerGlobalScope/DedicatedWorkerGlobalScope to stop
processing events and shut down. A closing flag was added to
WorkerGlobalScope per the spec.

Additionally, a closing flag was added to dom::worker::Worker to ensure
no further events are dispatched to the onmessage handler.

The change to the wpt WorkerTerminate.js is a patch for the broken
Worker_terminate_event_queue test (asserts 10001 == 10000).

Resolves servo/servo#4427

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

jwilm commented on components/script/dom/worker.rs:
```diff
@@ -100,6 +103,10 @@ impl Worker {
                           data: StructuredCloneData) {
         let worker = address.to_temporary().root();
 
+        if worker.r().closing.get() {
```
[The spec](https://html.spec.whatwg.org/multipage/#terminate-a-worker) specifies to empty the event/task queue associated with the worker as to not deliver further events to the `onmessage` handler. Without a synchronous way to affect the `DedicatedWorkerGlobalScope`, I resorted to these short-circuiting checks to prevent messages from being delivered. I keep thinking about this because it's kind of ugly. Does someone have any perspective on a cleaner way to achieve that behavior?

jwilm replied:
Gecko appears to differentiate between normal events in the worker thread's queue and what it calls a ControlRunnable. The run loop in their worker prioritizes the control events and handles at most one normal event on an iteration. We could implement something similar by adding a control channel and moving to a select model for the loop.

Ms2ger commented:
I commented on critic.

jwilm commented:
I left a couple of comments/questions in reply.

Do reviewers get notified of replies from critic, or should I ping from the issue?

jdm commented:
The comments arrive in email form; no ping necessary.

jwilm committed:
fixup! Implement Worker.prototype.terminate()

jwilm commented:
I fixed all of the issues on critic except escaping a long running JS handler. Interrupting JS execution in this case requires a call to JS_RequestInterrupt from `Worker`. This means that both `Worker` and `DedicatedWorkerGlobalScope` need a reference to the `Runtime`. However, we cannot Send `Runtime` or `Cx` in the current rust-mozjs, and so there's no way to schedule an interrupt. A higher level API in rust-mozjs that handles safe concurrent access to the runtime would make this possible.

jwilm committed:
fixup! Implement Worker.prototype.terminate()

jwilm commented:
The latest changes add support for killing a blocking script in the worker. Such capability requires a reference to the JSRuntime in `Worker`. I think this highlights the need for building out a sendable runtime in `mozjs-rust`. There's a couple of unsafe sections in both DedicatedWorkerGlobalScope and Worker to handle managing the mozjs interrupts. Again, we need more safe APIs from mozjs-rust.

It might be good to have a second channel for workers as well. I ended up needing to buffer ScriptMsgs outside of the receiver. This wouldn't even be a problem if there was a dedicated channel for control messages.

The code in `tests/html/worker` was for validation that the thread count was not increasing after spawning/terminating numerous workers. It would be nice to automate that, but for now there's test pages.

jwilm committed:
fixup! Implement Worker.prototype.terminate()

jwilm committed:
fixup! Implement Worker.prototype.terminate()

jwilm committed:
fixup! Implement Worker.prototype.terminate()

jwilm committed:
Refactor shared code for global object unwrapping

global_object_for_js_object and global_object_for_js_context shared a
lot of common code that has been moved into global_object_for_js_global.

Ms2ger commented:
Please rebase on master.

Ms2ger commented:
Rebased as https://github.com/servo/servo/pull/6652
